fix: parse forked bazel versions

PeterCardenas opened the pull request:
## Problem

Was doing some more work with bazel and was curious whether this lsp had updates. I was working with a forked version of bazel and hard coded the versions array to include mine. This seemed to cause the build to fail after updating.

The root cause was that version included a `/` to signal to `bazelisk` that it should pull the bazel version from a github repo. The caused the build to fail due to a parent directory of the flags cache not existing.

## Solution

I added a check to ensure that the parent directory of the flags cache exists before attempting to create/write to it. Also, added the error message in the build step for easier future debugging.

vogelsgesang commented:
Code changes look good, but I am not sure they are sufficient.

Do you have something like `/my-fork/7.5.1` in your `.bazelversion` / `USE_BAZEL_VERSION` variable?

I don't think the mechanism in [find_closest_bazel_version](https://github.com/salesforce-misc/bazelrc-lsp/blob/8705d2b5ee2e902e28068f93c1b25bee061df00b/src/bazel_version.rs#L29) could handle that, can it?

PeterCardenas commented:
Updated the parsing logic to account for forked bazel versions as well and updated tests. I'm a novice at rust so let me know if I should do anything differently 😅

PeterCardenas commented:
also any thoughts on adding a pre-push hook? would like to avoid eating up ci time but also prefer to keep auto format off

vogelsgesang commented on src/bazel_version.rs:
```diff
@@ -3,16 +3,56 @@ use once_cell::sync::Lazy;
+        }
+        None => (None, full_version_str),
+    };
+    let (mut parts, pre_release) = match version_str.find("-") {
```
afaik, `split_once` instead of `find` would be more concise?

vogelsgesang commented on src/bazel_version.rs:
```diff
@@ -84,13 +130,76 @@ pub fn auto_detect_bazel_version() -> Option<(String, Option<String>)> {
+        })
+    );
+    assert_eq!(
+        parse_bazel_version("9.0.0-pre.20210317.1"),
```
please also add a test case containing a fork

vogelsgesang commented:
By the way: Not sure what your exact use case is, but maybe https://github.com/salesforce-misc/bazelrc-lsp/commit/6dec094181c6104c736589ae85a0cfecc9bb2ee2 is also helpful for you

PeterCardenas commented:
oh haha i thought i'd left a comment on that commit. it's definitely helpful! there were some issues with the patch i had to include the forked bazel flags that this auto magically resolved so no need for the patch! i was wondering if you were keen on detecting if bazelisk exists and then using your functionality automatically. i can try adding a commit if so :)

vogelsgesang commented:
> that this auto magically resolved so no need for the patch!

I think I will still apply parts of the patch, in particular the better error reporting and the parsing of pre-release versions seems generally useful. For the "forked support", I am honestly still not sure how it works. I never used bazelisk with forked Bazel versions. Can you point me to some documentation / code which show me how bazelisk behaves with forked versions?

> i was wondering if you were keen on detecting if bazelisk exists and then using your functionality automatically. i can try adding a commit if so :)

I didn't quite wrap my head around the security implications of running `bazelisk help flags-as-proto` on a potentially untrusted repository.

Depending on the contents of the `.bazelversion`/`.bazeliskrc`, it might download an arbitrary binary from the internet and execute it. That binary might not even be Bazel, from what I can tell, and could do arbitrary harm. While the user would also harm themselves by running `bazelisk` manually in such a repository, at least I don't want to cause that harm automatically via this LSP.

I know that at least VS-Code has a differentiation between "trusted" and "untrusted" projects, but I didn't find an equivalent in the LSP world. I think that having auto-detection for bazelisk would be valuable, but it needs to be behind some kind of opt-in / confirmation.

PeterCardenas commented:
ah i was referring that there's no need for my hacky patch since i can just use yours 😅 
on the security part, i think that's a fair assessment. maybe what can be done instead is we can auto install any version that isn't a fork (pre-release also seems fine but can also err on the side of caution there). We can also add an lsp setting that the user will have to explicitly enable certain forked versions of bazel, since having them all allowed is good for productivity/minimal configuration, but is likely not worth it in terms of security 🙃 right now how i have it is that i just set the env var to bazelisk which is less secure than the specific lsp setting 😅
